src/metabase/lib/metadata/calculation.cljc

```diff
@@ -91,6 +91,33 @@
                          :stage-number stage-number}
                         e)))))))
 
+(defmulti field-id-method
+  "Find the field id for something or nil."
+  {:arglists '([query stage-number x])}
+  (fn [_query _stage-number x]
+    (lib.dispatch/dispatch-value x))
+  :hierarchy lib.hierarchy/hierarchy)
+
+(defmethod field-id-method :default
+  [_query _stage-number _x]
+  nil)
+
+(mu/defn field-id :- [:maybe ::lib.schema.common/int-greater-than-or-equal-to-zero]
```

Comment:
I'm not really sure if this is something I would consider to be metadata "calculation", maybe this should just go in `metabase.lib.field` instead. Then you can make `field-id-method` `^:private`. It's also consistent with `metabase.lib.database/database-id`